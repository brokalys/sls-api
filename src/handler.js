import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
  ApolloServerPluginUsageReporting,
  ApolloServerPluginUsageReportingDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';
const isTestMode = process.env.NODE_ENV === 'test';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (isDevMode && !isTestMode) ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (!isTestMode) {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    !isDevMode && !isTestMode
      ? ApolloServerPluginUsageReporting({
          sendVariableValues: {
            exceptNames: ['email', 'unsubscribe_key'],
          },
          logger: {
            debug: () => {},
            info: () => {},
            warn: console.warn,
            error: console.error,
          },
          generateClientInfo: ({ context, request }) => {
            const headers = request.http && request.http.headers;
            const clientVersion =
              headers['apollographql-client-version'] || '1.0.0';

            if (!context.user) {
              return {
                clientName: 'Unauthorized',
                clientVersion,
              };
            }

            return {
              clientName: context.user.apiKey.name,
              clientVersion,
            };
          },
        })
      : ApolloServerPluginUsageReportingDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
