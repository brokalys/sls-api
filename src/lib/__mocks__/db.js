export default {
  connect: jest.fn(),
  getClient: jest.fn(() => ({
    escape: jest.fn(),
  })),
  query: jest.fn(() => []),
  end: jest.fn(),
};
