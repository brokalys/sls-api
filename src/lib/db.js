import serverlessMysql from 'serverless-mysql';

const mysql = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    database: process.env.DB_DATABASE,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
  },
  maxRetries: 5,
});

if (process.env.NODE_ENV !== 'test') {
  mysql.connect();
}

export default mysql;
