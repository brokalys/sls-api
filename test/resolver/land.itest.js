import { createTestClient } from 'apollo-server-testing';
import { server } from 'handler';
import { authenticateAs } from 'helpers';
import db from 'db-config';

const { query } = createTestClient(server);

describe('Resolver: land', () => {
  beforeAll(async () => {
    await db.migrate.latest();
    await db.seed.run();
  });

  afterEach(jest.clearAllMocks);

  test('successfully retrieves single land', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        query($id: Int!) {
          land(id: $id) {
            id
          }
        }
      `,
      variables: {
        id: 1,
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves all the available fields', async () => {
    authenticateAs('slsCrawler', server);

    const response = await query({
      query: `
        query($id: Int!) {
          land(id: $id) {
            id
            bounds
            cadastral_designation
            object_code
            area
            group_code
            location_classificator
            properties {
              results {
                id
                url
                category
                type
                rent_type
                content
                images
                price
                calc_price_per_sqm
                rooms
                area
                floor
                lat
                lng
                published_at
                foreign_id
              }
              summary {
                count
                price {
                  count
                  min
                  max
                  mean
                  median
                  mode
                  standardDev
                }
              }
            }
            vzd {
              land {
                id
                sale_id
                cadastre_number
                property_address
                sale_date
                price
                land_cadastral_designations
                land_total_area_m2
              }
            }
          }
        }
      `,
      variables: {
        id: 1,
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves land an property information', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        query($id: Int!) {
          land(id: $id) {
            id
            properties {
              results {
                price
              }
            }
          }
        }
      `,
      variables: {
        id: 1,
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves land VZD sales information', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        query($id: Int!) {
          land(id: $id) {
            id
            vzd {
              land {
                id
                cadastre_number
                price
              }
            }
          }
        }
      `,
      variables: {
        id: 1,
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('throws a validation exception if no id filter provided', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        {
          land {
            id
          }
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('throws an authentication error if trying to retrieve results without authorizing', async () => {
    const response = await query({
      query: `
        query($id: Int!) {
          land(id: $id) {
            id
            properties {
              results {
                price
              }
            }
          }
        }
      `,
      variables: {
        id: 1,
      },
    });

    expect(response.errors).toEqual([
      expect.objectContaining({
        extensions: {
          code: 'UNAUTHENTICATED',
        },
      }),
    ]);
  });

  test('throws an authentication error if trying to retrieve VZD land results without authorizing', async () => {
    const response = await query({
      query: `
        query($id: Int!) {
          land(id: $id) {
            id
            vzd {
              land {
                price
              }
            }
          }
        }
      `,
      variables: {
        id: 1,
      },
    });

    expect(response.errors).toEqual([
      expect.objectContaining({
        extensions: {
          code: 'UNAUTHENTICATED',
        },
      }),
    ]);
  });
});
