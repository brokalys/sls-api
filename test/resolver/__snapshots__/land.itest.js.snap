// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolver: land successfully retrieves all the available fields 1`] = `
Object {
  "data": Object {
    "land": Object {
      "area": 100,
      "bounds": "56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619",
      "cadastral_designation": "42010053310001",
      "group_code": "9894006",
      "id": 1,
      "location_classificator": "latvia-riga-centrs",
      "object_code": "5201011310",
      "properties": Object {
        "results": Array [
          Object {
            "area": 90,
            "calc_price_per_sqm": 1.11,
            "category": "apartment",
            "content": "",
            "floor": 2,
            "foreign_id": null,
            "id": 1,
            "images": Array [],
            "lat": 56.1,
            "lng": 24.9,
            "price": 100,
            "published_at": "2021-04-12T19:47:30.000Z",
            "rent_type": "monthly",
            "rooms": null,
            "type": "rent",
            "url": "https://www.ss.com/msg/lv/real-estate/flats/aizkraukle-and-reg/aiviekstes-pag/aghfx.html",
          },
          Object {
            "area": null,
            "calc_price_per_sqm": null,
            "category": "apartment",
            "content": "",
            "floor": null,
            "foreign_id": null,
            "id": 2,
            "images": Array [],
            "lat": 56.1,
            "lng": 24.9,
            "price": 200,
            "published_at": "2021-04-12T19:47:30.000Z",
            "rent_type": "monthly",
            "rooms": null,
            "type": "rent",
            "url": "https://www.ss.lv/msg/lv/real-estate/flats/riga/bolderaya/cceeb.html",
          },
        ],
        "summary": Object {
          "count": 2,
          "price": Object {
            "count": 2,
            "max": 200,
            "mean": 150,
            "median": 150,
            "min": 100,
            "mode": 100,
            "standardDev": 50,
          },
        },
      },
      "vzd": Object {
        "land": Array [
          Object {
            "cadastre_number": "76860060032",
            "id": 1,
            "land_cadastral_designations": Array [
              "76860060032",
            ],
            "land_total_area_m2": 479000,
            "price": 39000,
            "property_address": "\\"Bezdelīgactiņas\\", Turku pag., Līvānu nov., LV-5316",
            "sale_date": "2020-09-16T00:00:00.000Z",
            "sale_id": 1,
          },
        ],
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: land successfully retrieves land VZD sales information 1`] = `
Object {
  "data": Object {
    "land": Object {
      "id": 1,
      "vzd": Object {
        "land": Array [
          Object {
            "cadastre_number": "76860060032",
            "id": 1,
            "price": 39000,
          },
        ],
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: land successfully retrieves land an property information 1`] = `
Object {
  "data": Object {
    "land": Object {
      "id": 1,
      "properties": Object {
        "results": Array [
          Object {
            "price": 100,
          },
          Object {
            "price": 200,
          },
        ],
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: land successfully retrieves single land 1`] = `
Object {
  "data": Object {
    "land": Object {
      "id": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: land throws a validation exception if no id filter provided 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "GRAPHQL_VALIDATION_FAILED",
      },
      "message": "Field \\"land\\" argument \\"id\\" of type \\"Int!\\" is required, but it was not provided.",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
